library/src/main/java/com/novoda/downloadmanager/CallbackThrottleByProgressIncrease.java

```diff
@@ -3,40 +3,51 @@
 class CallbackThrottleByProgressIncrease implements CallbackThrottle {
 
     private DownloadBatchStatusCallback callback;
-    private DownloadBatchStatus downloadBatchStatus;
+
     private int currentProgress;
+    private DownloadBatchStatus.Status currentStatus;
+    private DownloadError currentDownloadError;
 
     @Override
     public void setCallback(DownloadBatchStatusCallback callback) {
         this.callback = callback;
     }
 
     @Override
-    public void update(DownloadBatchStatus downloadBatchStatus) {
+    public void update(DownloadBatchStatus currentDownloadBatchStatus) {
         if (callback == null) {
             return;
         }
 
-        this.downloadBatchStatus = downloadBatchStatus;
+        if (statusHasChanged(currentDownloadBatchStatus)
+                || progressHasChanged(currentDownloadBatchStatus)
+                || errorHasChanged(currentDownloadBatchStatus)) {
+
+            currentStatus = currentDownloadBatchStatus.status();
+            currentProgress = currentDownloadBatchStatus.percentageDownloaded();
+            currentDownloadError = currentDownloadBatchStatus.downloadError();
 
-        if (progressIncreased(downloadBatchStatus)) {
-            callback.onUpdate(downloadBatchStatus);
+            callback.onUpdate(currentDownloadBatchStatus);
         }
     }
 
-    private boolean progressIncreased(DownloadBatchStatus downloadBatchStatus) {
-        int newProgress = downloadBatchStatus.percentageDownloaded();
+    private boolean statusHasChanged(DownloadBatchStatus currentDownloadBatchStatus) {
+        DownloadBatchStatus.Status newStatus = currentDownloadBatchStatus.status();
+        return newStatus != null && !newStatus.equals(currentStatus);
+    }
+
+    private boolean progressHasChanged(DownloadBatchStatus currentDownloadBatchStatus) {
+        int newProgress = currentDownloadBatchStatus.percentageDownloaded();
+        return currentProgress != newProgress;
+    }
 
-        if (currentProgress == newProgress) {
-            return false;
-        } else {
-            currentProgress = newProgress;
-            return true;
-        }
+    private boolean errorHasChanged(DownloadBatchStatus currentDownloadBatchStatus) {
+        DownloadError newDownloadError = currentDownloadBatchStatus.downloadError();
+        return newDownloadError != null && !newDownloadError.equals(currentDownloadError);
     }
 
     @Override
     public void stopUpdates() {
-        callback.onUpdate(downloadBatchStatus);
+        callback = null;
```

Comment:
Are you sure about this? are we not risking that the last status is not emitted?

Reply:
I ran through the app and this doesn't seem to cause any issues, it was strange that we were saying to stop updates and yet it was spitting out an additional anyway.

Reply:
Ok, here's a scenario that illustrates why it's useful to send another update:
- file is 100 MB
- the previous update forwarded to the client was at 99.000 KBytes downloaded (99%)
- we download 999 kBytes, but given that the percentage didn't change (99%) we don't send a new update to the client
- download stops (for any reason, say network issue)
- `stopUpdates()` is invoked
- the client will never know that we actually download almost another 1MB (1%) of the file

Now the numbers might not be entirely allright in my example, but I hope I managed to explain why we need to send one last update here (maybe with some additional logic, about the downloaded data?)

Reply:
The important part from this is `download stops (for any reason, say network issue)` this will be notified because there was a state change, then the `stopUpdates` is invoked.

Reply:
Mmm good point 👍  That should cover all the edge cases where before we needed to send this extra update